File: 2021/cpp/day_5.cpp
//
// Created by trox667 on 05.12.21.
//

#include "day_5.hpp"
#include <tuple>
#include <cassert>
#include "utils.hpp"
#include <map>
#include <numeric>

namespace day5 {

    using Point = std::tuple<int, int>;

    struct Line {
        Line(Point &p1, Point &p2) : p1{std::move(p1)}, p2{std::move(p2)} {}

        [[nodiscard]] std::vector<Point> range(bool diagonal = false) const {
            auto result = std::vector<Point>{};
            auto ix = std::get<0>(p1) < std::get<0>(p2) ? 1 : -1;
            auto iy = std::get<1>(p1) < std::get<1>(p2) ? 1 : -1;
            auto dx = std::abs(std::get<0>(p1) - std::get<0>(p2));
            auto dy = std::abs(std::get<1>(p1) - std::get<1>(p2));
            auto c = std::max(dx, dy);

            if (dx != 0 && dy != 0 && !diagonal) { return result; }

            auto x = std::get<0>(p1);
            auto y = std::get<1>(p1);

            result.emplace_back(Point{x, y});

            for (auto i = 0; i < c; ++i) {
                if (dx != 0) x += ix;
                if (dy != 0) y += iy;
                result.emplace_back(Point{x, y});
            }

            return result;
        }

        Point p1;
        Point p2;
    };

    [[nodiscard]] std::vector<Line>
    create_lines(const std::vector<std::string> &input_lines) {
        auto lines = std::vector<Line>{};
        for (const auto &input_line: input_lines) {
            auto pos = input_line.find("->");
            if (pos != std::string::npos) {
                auto tmp = input_line;
                tmp.replace(pos, 2, ",");
                auto tokens = utils::split(tmp, ',');
                auto numbers = std::vector<int>(tokens.size());
                std::transform(tokens.begin(), tokens.end(), numbers.begin(),
                               [](const auto &token) {
                                   return std::stoi(token);
                               });
                assert(numbers.size() == 4);
                auto p1 = Point{numbers[0], numbers[1]};
                auto p2 = Point{numbers[2], numbers[3]};
                lines.emplace_back(Line(p1, p2));
            }

        }
        return lines;
    }

    [[nodiscard]] int
    run(const std::vector<std::string> &lines, bool diagonal = false) {
        auto points = std::map<Point, int>{};
        for (const auto &line: create_lines(lines)) {
            for (const auto &point: line.range(diagonal)) {
                points[point] += 1;
            }
        }
        int sum = 0;
        for (const auto &item: points) {
            if (item.second >= 2) sum++;
        }

        return sum;
    }


    [[nodiscard]] int part1(const std::vector<std::string> &lines) {
        return run(lines);
    }

    [[nodiscard]] int part2(const std::vector<std::string> &lines) {
        return run(lines, true);
    }
}

File: 2021/python/day5.py
import re
from collections import defaultdict


class Line:
    def __init__(self, values):
        self.x1 = values[0]
        self.y1 = values[1]
        self.x2 = values[2]
        self.y2 = values[3]

    def range(self):
        x_incr = 1 if self.x1 <= self.x2 else -1
        y_incr = 1 if self.y1 <= self.y2 else -1
        return range(self.x1, self.x2 + x_incr, x_incr), range(self.y1,
                                                               self.y2 + y_incr,
                                                               y_incr)

    def is_horizontal(self):
        return self.y1 == self.y2

    def is_vertical(self):
        return self.x1 == self.x2

    def is_diagonal(self):
        return abs(self.x1 - self.x2) == abs(self.y1 - self.y2)


def line_generator(lines, diagonal):
    for line in lines:
        x_range, y_range = line.range()
        if line.is_horizontal():
            for x in x_range:
                yield x, line.y1
        elif line.is_vertical():
            for y in y_range:
                yield line.x1, y
        elif diagonal and line.is_diagonal():
            for x, y in zip(x_range, y_range):
                yield x, y


def load_file():
    with open('../inputs/input05') as file:
        return file.read()


def file_to_lines():
    return [Line(list(map(int, line.replace('->', ',').split(',')))) for line
            in
            load_file().splitlines()]


def run(diagonal=False):
    visited = defaultdict(int)
    for coord in line_generator(file_to_lines(), diagonal):
        visited[coord] += 1
    return sum([1 for match in visited.values() if match > 1])


def part1():
    return run()


def part2():
    return run(True)


print(part1())
print(part2())
